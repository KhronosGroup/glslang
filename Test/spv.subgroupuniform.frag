#version 450

#extension GL_EXT_subgroupuniform_qualifier : enable

layout(location=0) subgroupuniformEXT in vec4 su_inv4;
subgroupuniformEXT float su_gf;
layout(location=1) in subgroupuniformEXT flat int su_ii;

layout(binding=0, input_attachment_index = 0) uniform subpassInput        inputAttachmentDyn[];
layout(binding=1)                             uniform samplerBuffer       uniformTexelBufferDyn[];
layout(binding=2, r32f)                       uniform imageBuffer         storageTexelBufferDyn[];
layout(binding=3)                             uniform uname { float a; }  uniformBuffer[];
layout(binding=4)                             buffer  bname { float b; }  storageBuffer[];
layout(binding=5)                             uniform sampler2D           sampledImage[];
layout(binding=6, r32f)                       uniform image2D             storageImage[];
layout(binding=7, input_attachment_index = 1) uniform subpassInput        inputAttachment[];
layout(binding=8)                             uniform samplerBuffer       uniformTexelBuffer[];
layout(binding=9, r32f)                       uniform imageBuffer         storageTexelBuffer[];

subgroupuniformEXT int foo(subgroupuniformEXT int nupi, subgroupuniformEXT out int f)
{
    return nupi;
}

void main()
{
    subgroupuniformEXT int su_li;
    int dyn_i;

    int a = foo(su_li, su_li);
    su_li = subgroupuniformEXT(a) + subgroupuniformEXT(a * 2);

    float b;
    b = su_inv4.x * su_gf;

    subgroupuniformEXT vec4 v;
    subgroupuniformEXT mat4 m;
    subgroupuniformEXT struct S { int a; } s;
    ivec4 uv;

    vec4 implicitlySubgroupuniform = m * v;

    // Although in SPIR-V the above's SSA node will be implicitly subgroup
    // uniform, because its type is not qualified it won't propagate its
    // uniformity further
    vec4 loseImplicitlySubgroupuniform = implicitlySubgroupuniform * 4.0f;

    // Check there is a warning for initializing something with unknown
    // uniformity to a subgroup uniform variable
    subgroupuniformEXT vec4 gainingUniformity = loseImplicitlySubgroupuniform;

    // Check that there is a warning on assignment too
    gainingUniformity = loseImplicitlySubgroupuniform;

    // But there is no warning for initializing a constant
    subgroupuniformEXT vec4 initializingFromConstant = vec4(4.0f);

    // Check there is no warning on assignment too
    initializingFromConstant = vec4(4.0f);
}
