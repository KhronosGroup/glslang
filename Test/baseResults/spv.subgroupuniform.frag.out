spv.subgroupuniform.frag
WARNING: 0:50: '=' :  subgroupuniform variable is initialized with a value of unknown uniformity
WARNING: 0:53: '=' :  subgroupuniform variable is assigned with a value of unknown uniformity

// Module Version 10000
// Generated by (magic number): 80007
// Id's are bound by 96

                              Capability Shader
                              Capability InputAttachment
                              Capability SampledBuffer
                              Capability ImageBuffer
               1:             ExtInstImport  "GLSL.std.450"
                              MemoryModel Logical GLSL450
                              EntryPoint Fragment 4  "main" 33 61
                              ExecutionMode 4 OriginUpperLeft
                              Source GLSL 450
                              SourceExtension  "GL_EXT_subgroupuniform_qualifier"
                              Name 4  "main"
                              Name 11  "foo(i1;i1;"
                              Name 9  "nupi"
                              Name 10  "f"
                              Name 16  "a"
                              Name 17  "su_li"
                              Name 18  "param"
                              Name 20  "param"
                              Name 30  "b"
                              Name 33  "su_inv4"
                              Name 39  "su_gf"
                              Name 43  "implicitlySubgroupuniform"
                              Name 46  "m"
                              Name 48  "v"
                              Name 51  "loseImplicitlySubgroupuniform"
                              Name 55  "gainingUniformity"
                              Name 58  "initializingFromConstant"
                              Name 61  "su_ii"
                              Name 66  "inputAttachmentDyn"
                              Name 71  "uniformTexelBufferDyn"
                              Name 75  "storageTexelBufferDyn"
                              Name 76  "uname"
                              MemberName 76(uname) 0  "a"
                              Name 79  "uniformBuffer"
                              Name 80  "bname"
                              MemberName 80(bname) 0  "b"
                              Name 83  "storageBuffer"
                              Name 88  "sampledImage"
                              Name 92  "storageImage"
                              Name 93  "inputAttachment"
                              Name 94  "uniformTexelBuffer"
                              Name 95  "storageTexelBuffer"
                              Decorate 13 Uniform
                              Decorate 17(su_li) Uniform
                              Decorate 19 Uniform
                              Decorate 23 Uniform
                              Decorate 26 Uniform
                              Decorate 27 Uniform
                              Decorate 33(su_inv4) Location 0
                              Decorate 33(su_inv4) Uniform
                              Decorate 38 Uniform
                              Decorate 39(su_gf) Uniform
                              Decorate 40 Uniform
                              Decorate 41 Uniform
                              Decorate 46(m) Uniform
                              Decorate 47 Uniform
                              Decorate 48(v) Uniform
                              Decorate 49 Uniform
                              Decorate 50 Uniform
                              Decorate 55(gainingUniformity) Uniform
                              Decorate 58(initializingFromConstant) Uniform
                              Decorate 61(su_ii) Flat
                              Decorate 61(su_ii) Location 1
                              Decorate 61(su_ii) Uniform
                              Decorate 66(inputAttachmentDyn) DescriptorSet 0
                              Decorate 66(inputAttachmentDyn) Binding 0
                              Decorate 66(inputAttachmentDyn) InputAttachmentIndex 0
                              Decorate 71(uniformTexelBufferDyn) DescriptorSet 0
                              Decorate 71(uniformTexelBufferDyn) Binding 1
                              Decorate 75(storageTexelBufferDyn) DescriptorSet 0
                              Decorate 75(storageTexelBufferDyn) Binding 2
                              MemberDecorate 76(uname) 0 Offset 0
                              Decorate 76(uname) Block
                              Decorate 79(uniformBuffer) DescriptorSet 0
                              Decorate 79(uniformBuffer) Binding 3
                              MemberDecorate 80(bname) 0 Offset 0
                              Decorate 80(bname) BufferBlock
                              Decorate 83(storageBuffer) DescriptorSet 0
                              Decorate 83(storageBuffer) Binding 4
                              Decorate 88(sampledImage) DescriptorSet 0
                              Decorate 88(sampledImage) Binding 5
                              Decorate 92(storageImage) DescriptorSet 0
                              Decorate 92(storageImage) Binding 6
                              Decorate 93(inputAttachment) DescriptorSet 0
                              Decorate 93(inputAttachment) Binding 7
                              Decorate 93(inputAttachment) InputAttachmentIndex 1
                              Decorate 94(uniformTexelBuffer) DescriptorSet 0
                              Decorate 94(uniformTexelBuffer) Binding 8
                              Decorate 95(storageTexelBuffer) DescriptorSet 0
                              Decorate 95(storageTexelBuffer) Binding 9
               2:             TypeVoid
               3:             TypeFunction 2
               6:             TypeInt 32 1
               7:             TypePointer Function 6(int)
               8:             TypeFunction 6(int) 7(ptr) 7(ptr)
              25:      6(int) Constant 2
              28:             TypeFloat 32
              29:             TypePointer Function 28(float)
              31:             TypeVector 28(float) 4
              32:             TypePointer Input 31(fvec4)
     33(su_inv4):     32(ptr) Variable Input
              34:             TypeInt 32 0
              35:     34(int) Constant 0
              36:             TypePointer Input 28(float)
              42:             TypePointer Function 31(fvec4)
              44:             TypeMatrix 31(fvec4) 4
              45:             TypePointer Function 44
              53:   28(float) Constant 1082130432
              59:   31(fvec4) ConstantComposite 53 53 53 53
              60:             TypePointer Input 6(int)
       61(su_ii):     60(ptr) Variable Input
              62:             TypeImage 28(float) SubpassData nonsampled format:Unknown
              63:     34(int) Constant 1
              64:             TypeArray 62 63
              65:             TypePointer UniformConstant 64
66(inputAttachmentDyn):     65(ptr) Variable UniformConstant
              67:             TypeImage 28(float) Buffer sampled format:Unknown
              68:             TypeSampledImage 67
              69:             TypeArray 68 63
              70:             TypePointer UniformConstant 69
71(uniformTexelBufferDyn):     70(ptr) Variable UniformConstant
              72:             TypeImage 28(float) Buffer nonsampled format:R32f
              73:             TypeArray 72 63
              74:             TypePointer UniformConstant 73
75(storageTexelBufferDyn):     74(ptr) Variable UniformConstant
       76(uname):             TypeStruct 28(float)
              77:             TypeArray 76(uname) 63
              78:             TypePointer Uniform 77
79(uniformBuffer):     78(ptr) Variable Uniform
       80(bname):             TypeStruct 28(float)
              81:             TypeArray 80(bname) 63
              82:             TypePointer Uniform 81
83(storageBuffer):     82(ptr) Variable Uniform
              84:             TypeImage 28(float) 2D sampled format:Unknown
              85:             TypeSampledImage 84
              86:             TypeArray 85 63
              87:             TypePointer UniformConstant 86
88(sampledImage):     87(ptr) Variable UniformConstant
              89:             TypeImage 28(float) 2D nonsampled format:R32f
              90:             TypeArray 89 63
              91:             TypePointer UniformConstant 90
92(storageImage):     91(ptr) Variable UniformConstant
93(inputAttachment):     65(ptr) Variable UniformConstant
94(uniformTexelBuffer):     70(ptr) Variable UniformConstant
95(storageTexelBuffer):     74(ptr) Variable UniformConstant
         4(main):           2 Function None 3
               5:             Label
           16(a):      7(ptr) Variable Function
       17(su_li):      7(ptr) Variable Function
       18(param):      7(ptr) Variable Function
       20(param):      7(ptr) Variable Function
           30(b):     29(ptr) Variable Function
       39(su_gf):     29(ptr) Variable Function
43(implicitlySubgroupuniform):     42(ptr) Variable Function
           46(m):     45(ptr) Variable Function
           48(v):     42(ptr) Variable Function
51(loseImplicitlySubgroupuniform):     42(ptr) Variable Function
55(gainingUniformity):     42(ptr) Variable Function
58(initializingFromConstant):     42(ptr) Variable Function
              19:      6(int) Load 17(su_li)
                              Store 18(param) 19
              21:      6(int) FunctionCall 11(foo(i1;i1;) 18(param) 20(param)
              22:      6(int) Load 20(param)
                              Store 17(su_li) 22
                              Store 16(a) 21
              23:      6(int) Load 16(a)
              24:      6(int) Load 16(a)
              26:      6(int) IMul 24 25
              27:      6(int) IAdd 23 26
                              Store 17(su_li) 27
              37:     36(ptr) AccessChain 33(su_inv4) 35
              38:   28(float) Load 37
              40:   28(float) Load 39(su_gf)
              41:   28(float) FMul 38 40
                              Store 30(b) 41
              47:          44 Load 46(m)
              49:   31(fvec4) Load 48(v)
              50:   31(fvec4) MatrixTimesVector 47 49
                              Store 43(implicitlySubgroupuniform) 50
              52:   31(fvec4) Load 43(implicitlySubgroupuniform)
              54:   31(fvec4) VectorTimesScalar 52 53
                              Store 51(loseImplicitlySubgroupuniform) 54
              56:   31(fvec4) Load 51(loseImplicitlySubgroupuniform)
                              Store 55(gainingUniformity) 56
              57:   31(fvec4) Load 51(loseImplicitlySubgroupuniform)
                              Store 55(gainingUniformity) 57
                              Store 58(initializingFromConstant) 59
                              Store 58(initializingFromConstant) 59
                              Return
                              FunctionEnd
  11(foo(i1;i1;):      6(int) Function None 8
         9(nupi):      7(ptr) FunctionParameter
           10(f):      7(ptr) FunctionParameter
              12:             Label
              13:      6(int) Load 9(nupi)
                              ReturnValue 13
                              FunctionEnd
